File: Go/luhn_utils/luhn_utils.go
// Package luhn_utils provides a comprehensive implementation of the Luhn algorithm
// (also known as the "modulus 10" or "mod 10" algorithm) for validating and
// generating check digits.
//
// The Luhn algorithm is used for validating:
//   - Credit card numbers
//   - IMEI numbers (International Mobile Equipment Identity)
//   - National Provider Identifier numbers (US healthcare)
//   - Canadian Social Insurance Numbers
//   - Greek Social Security Numbers (AMKA)
//   - South African ID numbers
//   - And many other identification numbers
//
// Reference: https://en.wikipedia.org/wiki/Luhn_algorithm
package luhn_utils

import (
	"errors"
	"math/rand"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Common card prefixes for identification
var cardPrefixes = map[string][]string{
	"visa":       {"4"},
	"mastercard": {"51", "52", "53", "54", "55", "2221", "2222", "2223", "2224", "2225"},
	"amex":       {"34", "37"},
	"discover":   {"6011", "622126", "644", "645", "646", "647", "648", "649", "65"},
	"jcb":        {"3528", "3529", "353", "354", "355", "356", "357", "358"},
	"diners":     {"300", "301", "302", "303", "304", "305", "36", "38", "39"},
	"unionpay":   {"62", "81"},
	"maestro":    {"5018", "5020", "5038", "5893", "6304", "6759", "6761", "6762", "6763"},
	"mir":        {"2200", "2201", "2202", "2203", "2204"},
}

// nonDigitRegex matches any non-digit character
var nonDigitRegex = regexp.MustCompile(`[^\d]`)

// Errors
var (
	ErrEmptyNumber    = errors.New("number must contain at least one digit")
	ErrInvalidLength  = errors.New("length must be greater than prefix length")
	ErrInvalidDigit   = errors.New("digit must be between 0 and 9")
	ErrInvalidNumber  = errors.New("number is not valid according to Luhn algorithm")
)

// StripFormatting removes all non-digit characters from a number string.
func StripFormatting(number string) string {
	return nonDigitRegex.ReplaceAllString(number, "")
}

// CalculateCheckDigit calculates the Luhn check digit for a given number.
// The check digit is the digit that, when appended to the number,
// makes it pass the Luhn validation.
//
// Returns the check digit (0-9) or an error if the number is invalid.
func CalculateCheckDigit(number string) (int, error) {
	cleanNumber := StripFormatting(number)
	
	if len(cleanNumber) == 0 {
		return 0, ErrEmptyNumber
	}
	
	total := 0
	digits := []rune(cleanNumber)
	
	// Double every other digit from right to left, starting with the first
	for i := len(digits) - 1; i >= 0; i-- {
		digit, err := strconv.Atoi(string(digits[i]))
		if err != nil {
			return 0, ErrInvalidDigit
		}
		
		position := len(digits) - 1 - i
		if position%2 == 0 {
			// Double this digit
			doubled := digit * 2
			if doubled > 9 {
				doubled -= 9
			}
			total += doubled
		} else {
			total += digit
		}
	}
	
	return (10 - (total % 10)) % 10, nil
}

// Validate validates a number using the Luhn algorithm.
// Returns true if the number is valid according to Luhn algorithm.
func Validate(number string) bool {
	cleanNumber := StripFormatting(number)
	
	if len(cleanNumber) < 2 {
		return false
	}
	
	total := 0
	digits := []rune(cleanNumber)
	
	// Double every second digit from right to left
	for i := len(digits) - 1; i >= 0; i-- {
		digit, err := strconv.Atoi(string(digits[i]))
		if err != nil {
			return false
		}
		
		position := len(digits) - 1 - i
		if position%2 == 1 {
			// Double this digit
			doubled := digit * 2
			if doubled > 9 {
				doubled -= 9
			}
			total += doubled
		} else {
			total += digit
		}
	}
	
	return total%10 == 0
}

// AddCheckDigit appends the Luhn check digit to a number.
// Returns the number with the check digit appended.
func AddCheckDigit(number string) (string, error) {
	checkDigit, err := CalculateCheckDigit(number)
	if err != nil {
		return "", err
	}
	cleanNumber := StripFormatting(number)
	return cleanNumber + strconv.Itoa(checkDigit), nil
}

// FormatNumber formats a number by grouping digits.
// groupSize is the number of digits per group (default: 4).
// separator is the separator between groups (default: space).
func FormatNumber(number string, groupSize int, separator string) string {
	cleanNumber := StripFormatting(number)
	
	if len(cleanNumber) == 0 {
		return ""
	}
	
	if groupSize <= 0 {
		groupSize = 4
	}
	if separator == "" {
		separator = " "
	}
	
	var groups []string
	for i := 0; i < len(cleanNumber); i += groupSize {
		end := i + groupSize
		if end > len(cleanNumber) {
			end = len(cleanNumber)
		}
		groups = append(groups, cleanNumber[i:end])
	}
	
	return strings.Join(groups, separator)
}

// GenerateValidNumber generates a valid Luhn number with the given prefix.
// This is useful for generating test credit card numbers.
//
// prefix is the prefix for the number (e.g., "4" for Visa).
// length is the desired total length of the number (default: 16).
func GenerateValidNumber(prefix string, length int) (string, error) {
	cleanPrefix := StripFormatting(prefix)
	
	if len(cleanPrefix) == 0 {
		return "", ErrEmptyNumber
	}
	
	if length <= len(cleanPrefix) {
		return "", ErrInvalidLength
	}
	
	// Initialize random source
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	
	// Fill remaining digits (except check digit) with random digits
	remainingLength := length - len(cleanPrefix) - 1
	randomDigits := make([]string, remainingLength)
	for i := 0; i < remainingLength; i++ {
		randomDigits[i] = strconv.Itoa(r.Intn(10))
	}
	
	// Combine prefix and random digits
	numberWithoutCheck := cleanPrefix + strings.Join(randomDigits, "")
	
	// Calculate and append check digit
	return AddCheckDigit(numberWithoutCheck)
}

// GenerateBatch generates multiple valid Luhn numbers with the given prefix.
func GenerateBatch(prefix string, count int, length int) ([]string, error) {
	results := make([]string, count)
	for i := 0; i < count; i++ {
		num, err := GenerateValidNumber(prefix, length)
		if err != nil {
			return nil, err
		}
		results[i] = num
	}
	return results, nil
}

// CardType represents a credit card type with its test numbers
type CardType struct {
	Type   string
	Number string
}

// GenerateTestCreditCards generates test credit card numbers for various card types.
// These are TEST numbers only and will not work for actual transactions.
// They are generated to pass Luhn validation.
func GenerateTestCreditCards(countPerType int) []CardType {
	// Initialize random source
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	
	var cards []CardType
	
	for cardType, prefixes := range cardPrefixes {
		for i := 0; i < countPerType; i++ {
			prefix := prefixes[r.Intn(len(prefixes))]
			
			// Use appropriate length based on card type
			length := 16
			switch cardType {
			case "amex":
				length = 15
			case "diners":
				length = 14
			}
			
			number, err := GenerateValidNumber(prefix, length)
			if err == nil {
				cards = append(cards, CardType{
					Type:   cardType,
					Number: number,
				})
			}
		}
	}
	
	return cards
}

// IdentifyCardType attempts to identify the card type based on the number prefix.
// This uses common prefix patterns and does not validate the number.
// Always validate the number separately before use.
func IdentifyCardType(number string) string {
	cleanNumber := StripFormatting(number)
	
	if len(cleanNumber) == 0 {
		return ""
	}
	
	for cardType, prefixes := range cardPrefixes {
		for _, prefix := range prefixes {
			if strings.HasPrefix(cleanNumber, prefix) {
				return cardType
			}
		}
	}
	
	return ""
}

// CalculateLuhnSum calculates the Luhn sum and returns both the sum and validity.
// This is useful for debugging or educational purposes.
func CalculateLuhnSum(number string) (int, bool) {
	cleanNumber := StripFormatting(number)
	
	if len(cleanNumber) == 0 {
		return 0, false
	}
	
	total := 0
	digits := []rune(cleanNumber)
	
	for i := len(digits) - 1; i >= 0; i-- {
		digit, err := strconv.Atoi(string(digits[i]))
		if err != nil {
			return 0, false
		}
		
		position := len(digits) - 1 - i
		if position%2 == 1 {
			doubled := digit * 2
			if doubled > 9 {
				doubled -= 9
			}
			total += doubled
		} else {
			total += digit
		}
	}
	
	return total, total%10 == 0
}

// FindCheckDigitErrors finds positions where a single digit error would make the number valid.
// This is useful for error detection and correction suggestions.
func FindCheckDigitErrors(number string) []int {
	cleanNumber := StripFormatting(number)
	
	if len(cleanNumber) == 0 {
		return nil
	}
	
	if Validate(cleanNumber) {
		return nil // Already valid
	}
	
	var errorPositions []int
	digits := []rune(cleanNumber)
	
	for i := 0; i < len(digits); i++ {
		original := digits[i]
		for newDigit := '0'; newDigit <= '9'; newDigit++ {
			if newDigit == original {
				continue
			}
			digits[i] = newDigit
			if Validate(string(digits)) {
				errorPositions = append(errorPositions, i)
				break
			}
		}
		digits[i] = original
	}
	
	return errorPositions
}

// Validator is a class-based validator for Luhn numbers.
// Provides a convenient interface for validating and generating Luhn-compliant numbers.
type Validator struct {
	groupSize int
	separator string
}

// NewValidator creates a new Validator with the given formatting options.
func NewValidator(groupSize int, separator string) *Validator {
	if groupSize <= 0 {
		groupSize = 4
	}
	if separator == "" {
		separator = " "
	}
	return &Validator{
		groupSize: groupSize,
		separator: separator,
	}
}

// Validate validates a number using the Luhn algorithm.
func (v *Validator) Validate(number string) bool {
	return Validate(number)
}

// CalculateCheckDigit calculates the check digit for a number.
func (v *Validator) CalculateCheckDigit(number string) (int, error) {
	return CalculateCheckDigit(number)
}

// AddCheckDigit appends the check digit to a number.
func (v *Validator) AddCheckDigit(number string) (string, error) {
	return AddCheckDigit(number)
}

// Format formats a number with the default group size and separator.
func (v *Validator) Format(number string) string {
	return FormatNumber(number, v.groupSize, v.separator)
}

// Strip removes formatting from a number.
func (v *Validator) Strip(number string) string {
	return StripFormatting(number)
}

// Generate generates a valid Luhn number with the given prefix.
func (v *Validator) Generate(prefix string, length int) (string, error) {
	return GenerateValidNumber(prefix, length)
}

// GenerateBatch generates multiple valid Luhn numbers.
func (v *Validator) GenerateBatch(prefix string, count int, length int) ([]string, error) {
	return GenerateBatch(prefix, count, length)
}

File: Zig/luhn_utils/src/luhn_utils.zig
const std = @import("std");

/// Luhn algorithm errors
pub const LuhnError = error{
    OutOfMemory,
    InvalidCharacter,
    EmptyInput,
    InvalidLength,
    CheckDigitMismatch,
};

/// Card type enumeration
pub const CardType = enum {
    visa,
    mastercard,
    amex,
    discover,
    diners_club,
    jcb,
    unionpay,
    maestro,
    mir,
    unknown,

    /// Get human-readable card name
    pub fn name(self: CardType) []const u8 {
        return switch (self) {
            .visa => "Visa",
            .mastercard => "Mastercard",
            .amex => "American Express",
            .discover => "Discover",
            .diners_club => "Diners Club",
            .jcb => "JCB",
            .unionpay => "UnionPay",
            .maestro => "Maestro",
            .mir => "Mir",
            .unknown => "Unknown",
        };
    }

    /// Check if card type is valid (not unknown)
    pub fn isValid(self: CardType) bool {
        return self != .unknown;
    }
};

/// Luhn validation result
pub const ValidationResult = struct {
    is_valid: bool,
    card_type: CardType,
    check_digit: u8,
    computed_check_digit: u8,
    formatted_number: []const u8,

    pub fn deinit(self: *ValidationResult, allocator: std.mem.Allocator) void {
        allocator.free(self.formatted_number);
    }
};

/// Password strength levels (for general password strength checking)
pub const StrengthLevel = enum {
    very_weak,
    weak,
    fair,
    good,
    strong,
    very_strong,

    pub fn score(self: StrengthLevel) u8 {
        return switch (self) {
            .very_weak => 0,
            .weak => 1,
            .fair => 2,
            .good => 3,
            .strong => 4,
            .very_strong => 5,
        };
    }

    pub fn fromScore(value: u8) StrengthLevel {
        return switch (value) {
            0 => .very_weak,
            1 => .weak,
            2 => .fair,
            3 => .good,
            4 => .strong,
            else => .very_strong,
        };
    }
};

// ============================================================================
// Core Luhn Algorithm Functions
// ============================================================================

/// Calculate the Luhn check digit for a number string (without the check digit)
/// Returns the check digit that would make the number valid
pub fn calculateCheckDigit(number: []const u8) LuhnError!u8 {
    if (number.len == 0) {
        return LuhnError.EmptyInput;
    }

    var sum: u32 = 0;
    var is_even = true; // Will be false for the rightmost digit after we process

    // Process from right to left
    var i: usize = number.len;
    while (i > 0) {
        i -= 1;
        const c = number[i];

        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }

        var digit: u32 = @as(u32, c - '0');

        if (is_even) {
            digit *= 2;
            if (digit > 9) {
                digit -= 9;
            }
        }

        sum += digit;
        is_even = !is_even;
    }

    // The check digit is (10 - (sum mod 10)) mod 10
    const check_digit: u8 = @as(u8, @intCast((10 - (sum % 10)) % 10));
    return check_digit;
}

/// Validate a number using the Luhn algorithm
/// Returns true if the number passes the Luhn check
pub fn validate(number: []const u8) LuhnError!bool {
    if (number.len == 0) {
        return LuhnError.EmptyInput;
    }

    var sum: u32 = 0;
    var is_even = false;

    // Process from right to left
    var i: usize = number.len;
    while (i > 0) {
        i -= 1;
        const c = number[i];

        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }

        var digit: u32 = @as(u32, c - '0');

        if (is_even) {
            digit *= 2;
            if (digit > 9) {
                digit -= 9;
            }
        }

        sum += digit;
        is_even = !is_even;
    }

    return sum % 10 == 0;
}

/// Validate with detailed result
pub fn validateWithDetails(allocator: std.mem.Allocator, number: []const u8) LuhnError!ValidationResult {
    if (number.len == 0) {
        return LuhnError.EmptyInput;
    }

    // Check for invalid characters
    for (number) |c| {
        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }
    }

    const is_valid = try validate(number);
    const card_type = detectCardType(number);

    // Extract check digit (last digit)
    const check_digit: u8 = number[number.len - 1] - '0';

    // Calculate what check digit should be
    const number_without_check = number[0 .. number.len - 1];
    const computed_check_digit = try calculateCheckDigit(number_without_check);

    // Format the number
    var formatted = std.ArrayList(u8).init(allocator);
    errdefer formatted.deinit();

    const format_config = getCardFormatConfig(card_type);
    var pos: usize = 0;
    var group_idx: usize = 0;

    for (number, 0..) |c, i| {
        if (group_idx < format_config.group_sizes.len) {
            if (pos == format_config.group_sizes[group_idx] and i > 0) {
                try formatted.append(' ');
                pos = 0;
                group_idx += 1;
            }
        }
        try formatted.append(c);
        pos += 1;
    }

    return ValidationResult{
        .is_valid = is_valid,
        .card_type = card_type,
        .check_digit = check_digit,
        .computed_check_digit = computed_check_digit,
        .formatted_number = try formatted.toOwnedSlice(),
    };
}

// ============================================================================
// Card Type Detection
// ============================================================================

const CardFormatConfig = struct {
    group_sizes: []const usize,
    expected_lengths: []const usize,
};

fn getCardFormatConfig(card_type: CardType) CardFormatConfig {
    return switch (card_type) {
        .visa => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 13, 16, 19 } },
        .mastercard => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 16 } },
        .amex => .{ .group_sizes = &.{ 4, 6, 5 }, .expected_lengths = &.{ 15 } },
        .discover => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 16, 19 } },
        .diners_club => .{ .group_sizes = &.{ 4, 6, 4 }, .expected_lengths = &.{ 14, 16, 19 } },
        .jcb => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 16, 19 } },
        .unionpay => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 16, 19 } },
        .maestro => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 12, 13, 14, 15, 16, 17, 18, 19 } },
        .mir => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{ 16 } },
        .unknown => .{ .group_sizes = &.{ 4, 4, 4, 4 }, .expected_lengths = &.{} },
    };
}

/// Detect card type from card number
pub fn detectCardType(number: []const u8) CardType {
    if (number.len < 2) return .unknown;

    // Get first digits for pattern matching
    const first_digit = number[0];
    const first_two = if (number.len >= 2) @as(u32, @as(u32, number[0] - '0') * 10 + @as(u32, number[1] - '0')) else 0;
    const first_three = if (number.len >= 3) @as(u32, @as(u32, number[0] - '0') * 100 + @as(u32, number[1] - '0') * 10 + @as(u32, number[2] - '0')) else 0;
    const first_four = if (number.len >= 4) @as(u32, @as(u32, number[0] - '0') * 1000 + @as(u32, number[1] - '0') * 100 + @as(u32, number[2] - '0') * 10 + @as(u32, number[3] - '0')) else 0;
    const first_six = if (number.len >= 6) @as(u32, @as(u32, number[0] - '0') * 100000 + @as(u32, number[1] - '0') * 10000 + @as(u32, number[2] - '0') * 1000 + @as(u32, number[3] - '0') * 100 + @as(u32, number[4] - '0') * 10 + @as(u32, number[5] - '0')) else 0;

    // Visa: starts with 4
    if (first_digit == '4') {
        if (number.len >= 13 and number.len <= 19) {
            return .visa;
        }
    }

    // Mastercard: 51-55 or 2221-2720
    if (first_two >= 51 and first_two <= 55) {
        return .mastercard;
    }
    if (first_four >= 2221 and first_four <= 2720) {
        return .mastercard;
    }

    // American Express: 34 or 37
    if (first_two == 34 or first_two == 37) {
        if (number.len == 15) {
            return .amex;
        }
    }

    // Discover: 6011, 644-649, 65, or 622126-622925
    if (first_four == 6011) return .discover;
    if (first_two == 65) return .discover;
    if (first_three >= 644 and first_three <= 649) return .discover;
    if (first_six >= 622126 and first_six <= 622925) return .discover;

    // Diners Club: 300-305, 36, 38, 39
    if (first_three >= 300 and first_three <= 305) return .diners_club;
    if (first_two == 36 or first_two == 38 or first_two == 39) return .diners_club;

    // JCB: 3528-3589
    if (first_four >= 3528 and first_four <= 3589) return .jcb;

    // UnionPay: 62
    if (first_two == 62) return .unionpay;

    // Maestro: 50, 56-69
    if (first_two == 50) return .maestro;
    if (first_two >= 56 and first_two <= 69) return .maestro;

    // Mir: 2200-2204
    if (first_four >= 2200 and first_four <= 2204) return .mir;

    return .unknown;
}

/// Check if number has valid length for its card type
pub fn hasValidLength(number: []const u8) bool {
    const card_type = detectCardType(number);
    const config = getCardFormatConfig(card_type);

    for (config.expected_lengths) |len| {
        if (number.len == len) {
            return true;
        }
    }

    return false;
}

// ============================================================================
// Utility Functions
// ============================================================================

/// Clean a number string (remove spaces and dashes)
pub fn cleanNumber(allocator: std.mem.Allocator, input: []const u8) LuhnError![]u8 {
    var result = std.ArrayList(u8).init(allocator);
    errdefer result.deinit();

    for (input) |c| {
        switch (c) {
            '0'...'9' => try result.append(c),
            ' ', '-' => {}, // Skip spaces and dashes
            else => return LuhnError.InvalidCharacter,
        }
    }

    if (result.items.len == 0) {
        return LuhnError.EmptyInput;
    }

    return result.toOwnedSlice();
}

/// Format a number with spaces for readability
pub fn formatNumber(allocator: std.mem.Allocator, number: []const u8, group_size: usize) LuhnError![]u8 {
    var result = std.ArrayList(u8).init(allocator);
    errdefer result.deinit();

    for (number, 0..) |c, i| {
        if (i > 0 and i % group_size == 0) {
            try result.append(' ');
        }
        try result.append(c);
    }

    return result.toOwnedSlice();
}

/// Generate a random valid Luhn number (for testing)
pub fn generateTestNumber(allocator: std.mem.Allocator, prefix: []const u8, length: usize) ![]u8 {
    if (length < prefix.len + 1) {
        return LuhnError.InvalidLength;
    }

    var result = std.ArrayList(u8).init(allocator);
    errdefer result.deinit();

    // Add prefix
    try result.appendSlice(prefix);

    // Fill with random digits (except last one)
    var rng = std.Random.DefaultPrng.init(@as(u64, @intCast(std.time.timestamp())));
    const random = rng.random();

    while (result.items.len < length - 1) {
        const digit = random.int(u8) % 10;
        try result.append('0' + digit);
    }

    // Calculate and append check digit
    const check_digit = try calculateCheckDigit(result.items);
    try result.append('0' + check_digit);

    return result.toOwnedSlice();
}

// ============================================================================
// IMEI Validation (uses Luhn)
// ============================================================================

/// Validate IMEI (International Mobile Equipment Identity)
/// IMEI is 15 digits, uses Luhn algorithm
pub fn validateIMEI(imei: []const u8) LuhnError!bool {
    if (imei.len != 15) {
        return LuhnError.InvalidLength;
    }

    // Check all digits
    for (imei) |c| {
        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }
    }

    return validate(imei);
}

/// Validate IMEISV (IMEI with Software Version)
/// IMEISV is 16 digits, does NOT use Luhn (last 2 digits are software version)
pub fn validateIMEISV(imeisv: []const u8) LuhnError!bool {
    if (imeisv.len != 16) {
        return LuhnError.InvalidLength;
    }

    // Check all digits
    for (imeisv) |c| {
        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }
    }

    // IMEISV first 14 digits + check digit should be valid IMEI
    // So we validate first 15 digits with Luhn
    return validate(imeisv[0..15]);
}

// ============================================================================
// National Identifier Validation (National ID numbers using Luhn)
// ============================================================================

/// Validate South African ID number (13 digits, uses Luhn)
pub fn validateSouthAfricanID(id: []const u8) LuhnError!bool {
    if (id.len != 13) {
        return LuhnError.InvalidLength;
    }

    for (id) |c| {
        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }
    }

    return validate(id);
}

/// Validate Canadian SIN (Social Insurance Number, 9 digits, uses Luhn)
pub fn validateCanadianSIN(sin: []const u8) LuhnError!bool {
    if (sin.len != 9) {
        return LuhnError.InvalidLength;
    }

    for (sin) |c| {
        if (c < '0' or c > '9') {
            return LuhnError.InvalidCharacter;
        }
    }

    return validate(sin);
}

// ============================================================================
// Tests
// ============================================================================

test "calculateCheckDigit" {
    // Test cases: number without check digit -> expected check digit
    try std.testing.expectEqual(@as(u8, 3), try calculateCheckDigit("7992739871"));
    try std.testing.expectEqual(@as(u8, 6), try calculateCheckDigit("4992739871"));
}

test "validate basic" {
    // Known valid numbers
    try std.testing.expect(try validate("79927398713"));
    try std.testing.expect(try validate("49927398716"));
    try std.testing.expect(try validate("4242424242424242")); // Visa test
    try std.testing.expect(try validate("5555555555554444")); // Mastercard test

    // Invalid numbers
    try std.testing.expect(!try validate("79927398710"));
    try std.testing.expect(!try validate("79927398711"));
    try std.testing.expect(!try validate("79927398712"));
    try std.testing.expect(!try validate("4242424242424241")); // Wrong check digit
}

test "validate with invalid character" {
    _ = validate("7992a398713") catch |err| {
        try std.testing.expectEqual(LuhnError.InvalidCharacter, err);
    };
}

test "validate empty" {
    _ = validate("") catch |err| {
        try std.testing.expectEqual(LuhnError.EmptyInput, err);
    };
}

test "detectCardType Visa" {
    try std.testing.expectEqual(CardType.visa, detectCardType("4242424242424242"));
    try std.testing.expectEqual(CardType.visa, detectCardType("4000000000000002"));
    try std.testing.expectEqual(CardType.visa, detectCardType("4111111111111111"));
}

test "detectCardType Mastercard" {
    try std.testing.expectEqual(CardType.mastercard, detectCardType("5555555555554444"));
    try std.testing.expectEqual(CardType.mastercard, detectCardType("2221000000000009"));
    try std.testing.expectEqual(CardType.mastercard, detectCardType("2720990000000006"));
}

test "detectCardType Amex" {
    try std.testing.expectEqual(CardType.amex, detectCardType("378282246310005"));
    try std.testing.expectEqual(CardType.amex, detectCardType("371449635398431"));
}

test "detectCardType Discover" {
    try std.testing.expectEqual(CardType.discover, detectCardType("6011111111111117"));
    try std.testing.expectEqual(CardType.discover, detectCardType("6011000990139424"));
}

test "detectCardType JCB" {
    try std.testing.expectEqual(CardType.jcb, detectCardType("3530111333300000"));
    try std.testing.expectEqual(CardType.jcb, detectCardType("3566002020360505"));
}

test "detectCardType Unknown" {
    try std.testing.expectEqual(CardType.unknown, detectCardType("1234567890123456"));
}

test "cleanNumber" {
    const allocator = std.testing.allocator;

    const result1 = try cleanNumber(allocator, "4242 4242 4242 4242");
    defer allocator.free(result1);
    try std.testing.expectEqualSlices(u8, "4242424242424242", result1);

    const result2 = try cleanNumber(allocator, "4242-4242-4242-4242");
    defer allocator.free(result2);
    try std.testing.expectEqualSlices(u8, "4242424242424242", result2);
}

test "formatNumber" {
    const allocator = std.testing.allocator;

    const result = try formatNumber(allocator, "4242424242424242", 4);
    defer allocator.free(result);
    try std.testing.expectEqualSlices(u8, "4242 4242 4242 4242", result);
}

test "validateIMEI" {
    // Valid IMEI (Luhn valid)
    try std.testing.expect(try validateIMEI("490154203237518"));

    // Invalid IMEI
    try std.testing.expect(!try validateIMEI("490154203237519"));

    // Wrong length
    _ = validateIMEI("49015420323751") catch |err| {
        try std.testing.expectEqual(LuhnError.InvalidLength, err);
    };
}

test "validateCanadianSIN" {
    // Valid Canadian SIN
    try std.testing.expect(try validateCanadianSIN("046454286"));

    // Invalid SIN
    try std.testing.expect(!try validateCanadianSIN("046454287"));
}

test "generateTestNumber" {
    const allocator = std.testing.allocator;

    const number = try generateTestNumber(allocator, "4242", 16);
    defer allocator.free(number);

    try std.testing.expectEqual(@as(usize, 16), number.len);
    try std.testing.expectEqualSlices(u8, "4242", number[0..4]);
    try std.testing.expect(try validate(number));
}

test "validateWithDetails" {
    const allocator = std.testing.allocator;

    var result = try validateWithDetails(allocator, "4242424242424242");
    defer result.deinit(allocator);

    try std.testing.expect(result.is_valid);
    try std.testing.expectEqual(CardType.visa, result.card_type);
}

test "CardType name" {
    try std.testing.expectEqualSlices(u8, "Visa", CardType.visa.name());
    try std.testing.expectEqualSlices(u8, "Mastercard", CardType.mastercard.name());
    try std.testing.expectEqualSlices(u8, "Unknown", CardType.unknown.name());
}

test "hasValidLength" {
    try std.testing.expect(hasValidLength("4242424242424242")); // Visa 16
    try std.testing.expect(hasValidLength("378282246310005")); // Amex 15
    try std.testing.expect(!hasValidLength("424242424")); // Too short
}